### PR TITLE
Add sed_type keyword to SpectralModel.plot

### PR DESCRIPTION
**Pull request by fabiopintore**
This PR solves the issue #3324. A `sed_type` keyword is added to `SpectralModel.plot` and it allows to plot the corresponding evaluation methods of the model.

TO DO: add the `sed_type` keyword to `SpectralModel.plot_error`

### CODE REVIEW

File: gammapy/modeling/models/spectral.py [hunk 1]
```diff
@@ -316,13 +319,35 @@ def plot(
 
         energy_min, energy_max = energy_range
         energy = MapAxis.from_energy_bounds(
-            energy_min, energy_max, n_points, energy_unit
-        ).edges
+                                            energy_min, energy_max, n_points, energy_unit
+                                            ).edges
+
+        if sed_type:
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:325]: The typical patter here would be:

```python3
if sed_type == "dnde":
   ...
elif sed_type == "flux":
    ...
elif sed_type == "eflux":
   ...
elif sed_type == "e2flux":
   ...
else:
    raise ValueError(f"Not a valid SED type {sed_type}")

```

File: gammapy/modeling/models/spectral.py [hunk 2]
```diff
@@ -316,13 +319,35 @@ def plot(
...
+            if sed_type == "dnde":
+                energy_power = 0
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:327]: The `energy_power` can remain an independent parameter, no need to define it here...just remove.

File: gammapy/modeling/models/spectral.py [hunk 3]
```diff
@@ -316,13 +319,35 @@ def plot(
...
+                energy = (energy[:-1] + energy[1:]) / 2.
+
+            if sed_type == "eflux":
+                energy_power = 0
+                flux = self.energy_flux(energy[:-1], energy[1:]).to("TeV cm-2 s-1")
+                energy = (energy[:-1] + energy[1:]) / 2.
+
+            y = self._plot_scale_flux(energy, flux, energy_power)
+
+        else:
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:346]: Once you change to the proposed patter above, this else block shouldn't be needed....

File: gammapy/modeling/models/spectral.py [hunk 4]
```diff
@@ -316,13 +319,35 @@ def plot(
...
+            # evaluate model
+            flux = self(energy).to(flux_unit)
 
-        y = self._plot_scale_flux(energy, flux, energy_power)
+            y = self._plot_scale_flux(energy, flux, energy_power)
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:350]: Remove indentation here...

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @fabiopintore, I have left some online comments.

File: gammapy/modeling/models/spectral.py [hunk 5]
```diff
@@ -273,6 +273,7 @@ def plot(
         energy_unit="TeV",
         flux_unit="cm-2 s-1 TeV-1",
         energy_power=0,
+        sed_type = "",
```
[the remark this answers is not in the document]
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:276]: OK.

File: gammapy/modeling/models/spectral.py [hunk 6]
```diff
@@ -302,6 +303,8 @@ def plot(
             Unit of the flux axis
         energy_power : int, optional
             Power of energy to multiply flux axis with
+        sed_type : str, optional
```
[the remark this answers is not in the document]
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:306]: OK.

File: gammapy/modeling/models/spectral.py (hunk 1 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:325]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 2 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:327]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 3 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:346]: OK.

File: gammapy/modeling/models/spectral.py (hunk 4 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:350]: OK.

File: gammapy/modeling/models/spectral.py [hunk 7]
```diff
@@ -316,11 +319,29 @@ def plot(
 
         energy_min, energy_max = energy_range
         energy = MapAxis.from_energy_bounds(
-            energy_min, energy_max, n_points, energy_unit
-        ).edges
+                                            energy_min, energy_max, n_points, energy_unit
+                                            ).edges
+
+        if sed_type:
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:325]: I think the `if sed_type` block has no effect, just remove...

File: gammapy/modeling/models/spectral.py [hunk 8]
```diff
@@ -316,11 +319,29 @@ def plot(
...
+            if sed_type == "dnde":
+                flux = self(energy).to(flux_unit)
+
+            elif sed_type == "e2dnde":
+                energy_power = 2
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:330]: I think the `energy_power` should remain independent, so just remove it here and handle the case "manually" like `energy ** 2 * self(energy).to(flux_unit)`

File: gammapy/modeling/models/spectral.py [hunk 9]
```diff
@@ -316,11 +319,29 @@ def plot(
...
+                flux = self(energy).to(flux_unit)
+
+            elif sed_type == "flux":
+                energy_power = 0
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:334]: Again remove the `energy_power` here...

File: gammapy/modeling/models/spectral.py [hunk 10]
```diff
@@ -316,11 +319,29 @@ def plot(
...
+                flux = self.integral(energy[:-1], energy[1:]).to("cm-2 s-1")
+                energy = (energy[:-1] + energy[1:]) / 2.
 
-        # evaluate model
-        flux = self(energy).to(flux_unit)
+            elif sed_type == "eflux":
+                energy_power = 0
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:339]: Remove `energy_power` here..

File: gammapy/modeling/models/spectral.py [hunk 11]
```diff
@@ -390,7 +414,27 @@ def plot_error(
             energy_min, energy_max, n_points, energy_unit
         ).edges
 
-        flux, flux_err = self.evaluate_error(energy).to(flux_unit)
+
+        if sed_type:
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:418]: Same as above...

File: gammapy/modeling/models/spectral.py [hunk 12]
```diff
@@ -390,7 +414,27 @@ def plot_error(
...
+            if sed_type == "dnde":
+                flux, flux_err = self.evaluate_error(energy).to(flux_unit)
+            
+            elif sed_type == "e2dnde":
+                energy_power = 2
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:423]: Remove and handle manually...

File: gammapy/modeling/models/spectral.py [hunk 13]
```diff
@@ -390,7 +414,27 @@ def plot_error(
...
+                flux, flux_err = self.evaluate_error(energy).to(flux_unit)
+
+            elif sed_type == "flux":
+                energy_power = 0
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:427]: Remove

File: gammapy/modeling/models/spectral.py [hunk 14]
```diff
@@ -390,7 +414,27 @@ def plot_error(
...
+                flux, flux_err = self.integral_error(energy[:-1], energy[1:]).to("cm-2 s-1")
+                energy = (energy[:-1] + energy[1:]) / 2.
+
+            elif sed_type == "eflux":
+                energy_power = 0
```
Comment (adonath) [inline gammapy/modeling/models/spectral.py:432]: Remove

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @fabiopintore, I have left a few more comments. In addition:
- The `energy_power` just remains independent, so even `e2dnde` can bis scaled again...
- I think it would be good to adapt the y axis label to the corresponding sed type

File: gammapy/modeling/models/spectral.py (hunk 7 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:325]: OK.

File: gammapy/modeling/models/spectral.py (hunk 8 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:330]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 9 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:334]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 10 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:339]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 11 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:418]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 12 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:423]: OK.

File: gammapy/modeling/models/spectral.py (hunk 14 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:432]: Ok.

File: gammapy/modeling/models/spectral.py (hunk 13 above)
Comment (fabiopintore) [reply gammapy/modeling/models/spectral.py:427]: Ok.

**Review by adonath — APPROVED** [review]
Thanks a lot @fabiopintore! No further comments demo my side..

**Pull request merged**